### ISSUE TITLE
[Feature Request] PEP 613 `typing.TypeAlias` support (...but officially deprecated, so may never happen)

### ISSUE THREAD

**Issue by jamesbraza** [beartype/beartype#328]
I was hoping to adopt `beartype` (version 0.17.2), but I found PEP 613's `typing.TypeAlias` is not yet supported. I am using Python 3.12:

```python
from typing import TypeAlias

MyID: TypeAlias = str
```

Running this with `beartype` will lead to this traceback:

```none
  File "/Users/user/code/repo/mypackage/a.py", line 3, in <module>
    MyID: TypeAlias = str
  File "/Users/user/code/repo/venv/lib/python3.12/site-packages/beartype/door/_doorcheck.py", line 97, in die_if_unbearable
    func_raiser = make_func_raiser(hint, conf)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ...
  File "/Users/user/code/repo/venv/lib/python3.12/site-packages/beartype/_check/code/codemake.py", line 706, in make_check_expr
    die_if_hint_pep_unsupported(
  File "/Users/user/code/repo/venv/lib/python3.12/site-packages/beartype/_util/hint/pep/utilpeptest.py", line 218, in die_if_hint_pep_unsupported
    raise BeartypeDecorHintPepUnsupportedException(
beartype.roar.BeartypeDecorHintPepUnsupportedException: Die_if_unbearable() or is_bearable() type hint typing.TypeAlias currently unsupported by @beartype.
```

The request is, can `beartype` support `TypeAlias`?

https://github.com/beartype/beartype/issues/297 is related, but it's for PEP695 type aliases, so this request is for explicit `TypeAlias`.

**Comment by leycec** [comment]
*...heh.* I was kinda wondering if PEP 613 was ever going to come up. Sadly, it came up a bit too late. PEP 695 (i.e., ``type`` aliases under Python ≥ 3.12), which @beartype more-or-less fully supports, **officially deprecates PEP 613:**

> We propose to deprecate the existing `typing.TypeAlias` introduced in [PEP 613](https://peps.python.org/pep-0613/). The new syntax eliminates its need entirely.

This puts @beartype in a bit of an awkward position. Technically, @beartype *could* now add support for obsolete deprecated PEPs like PEP 613. Pragmatically, is adding support for obsolete deprecated PEPs *really* a good idea or a sound investment of scarce open-source volunteer hours? Probably...

...**not.** You know? Instead, @beartype should just valiantly forge ahead with features everybody wants like deep type-checking of `dict` and `set` containers. Time is scarce, life is fleeting, and my face is a rictus grin. I definitely wish I had sufficient man-hours to hurl at literally every single PEP – even obsolete deprecated PEPs like this. I love supporting PEPs! This is fun for me. But I've only got so many keyboards and fingers.

Sucks, huh? I know. Your pain is my pain. Okay. It's not. You feel more pain about this than I do. Still, I am sad. :crying_cat_face:

**Comment by leycec** [comment]
*...heh.* I just remembered that Quebec bear bro @felixchenier cleverly invented a workaround under Python < 3.12. If you'd rather not refactor deprecated PEP 613 type aliases into supported PEP 695 `type` aliases, fear not! @beartype still supports your use case with PEP 484 type aliases everywhere – including under Python < 3.12.

Feast your weary eyes on this shadow madness:

```python
# This actually works. Hard to believe. Yet, it's all true. Probably. Maybe.
from beartype import beartype
from typing import NewType

# Type alias for a list of strings. Pretty cool, huh? Don't answer that.
list_of_strs = NewType('list_of_strs', list[str])

@beartype
def some_func(some_arg: list_of_strs) -> None:
    print(some_arg)

some_func(['This can't possibly work...', '...yet, it does!'])
```

...which prints:

```
['This can't possibly work...', '...yet, it does!']
```

In other words, @beartype already does everything you want – albeit not quite in the way that you want it to. Still, I party. :partying_face:

**Comment by jamesbraza** [comment]
Sounds good, yeah I guess `TypeAlias` will be deprecated over coming few years, one less import too.

One thing to note is `mypy==1.8` doesn't yet support PEP 695 type aliases: https://github.com/python/mypy/issues/15238

Feel free to close this out

**Comment by leycec** [comment]
**Wow. Wow.** @beartype actually got out ahead of mypy on something, huh? First time *that's* happened. Probably the last, too. Tonight, I celebrate by playing video games.

Thanks for being so understanding. Would you mind if I keep this open for a bit? If nothing else, @beartype should at least emit a `DeprecatingWarning` about `TypeAlias` being deprecated by PEP 695. Please accept this gold trophy for reminding me that I should do this. :trophy:

**Comment by jamesbraza** [comment]
Ha congratulations on beating `mypy`. I think a message instructing to use PEP 695 type aliases (and Python>=3.12) is fine.

If `beartype` supports PEP 695 type aliases, I am wondering if there's some way for `beartype` to pretend the PEP 613 `TypeAlias` had been a PEP 695 `type` alias from the start. In other words, it may be a minor lift for PEP 613 type alias to use PEP 695's back end representation and logic.

**Comment by leycec** [comment]
**...huh.** Didn't think of that one. That's super clever and possibly feasible. The issue, of course, is that sort of ruse would only work under Python ≥ 3.12. Under Python ≤ 3.11, behaviour becomes undefined. Still, I'm all for ruses.

Actually, we *can* do something under Python ≤ 3.11. It's a bit absurd, but the archaic PEP 484 `typing.NewType` factory is actually an undocumented technology for defining type aliases at runtime. @beartype fully supports it. So, @beartype could (in order):

1. Emit a `DeprecationWarning`. Even if we do this internal masquerade on everyone's behalf, users should still be informed that `TypeAlias` has been deprecated. Right? Deprecated things eventually get dragged behind the shed out back and disappeared into a hole in the ground.
1. Under:
   * Python ≥ 3.12, silently convert PEP 613 type aliases to PEP 695 `type` aliases. A joyful utopia is in sight.
   * Python ≤ 3.11, silently convert PEP 613 type aliases to PEP 484 `NewType`-style type aliases. This is horrible.

Now, let's see if @leycec does anything other than play video games. :rofl: :sob:

**Comment by leycec** [comment]
**Resolved by 6c2733e.** @beartype now graciously emits exactly one deprecation warning for each Python process that encounters a `typing.TypeAlias` type hint. Why one? Because the warning is insane. You do *not* want to see this warning more than once. Even seeing this warning once could make you question your already tenuous grip on Pythonic sanity:

```python
BeartypeDecorHintPep613DeprecationWarning: PEP 613 type hint
typing.TypeAlias deprecated by PEP 695. Consider either:
* Requiring Python >= 3.12 and refactoring PEP 613 type aliases into
  PEP 695 type aliases. Note that Python < 3.12 will hate you for
  this: e.g.,
    # Instead of this...
    from typing import TypeAlias
    alias_name: TypeAlias = alias_value

    # ..."just" do this.
    type alias_name = alias_value
* Refactoring PEP 613 type aliases into PEP 484 "typing.NewType"-based
  type aliases. Note that static type-checkers (e.g., mypy, pyright,
  Pyre) will hate you for this: e.g.,
    # Instead of this...
    from typing import TypeAlias
    alias_name: TypeAlias = alias_value

    # ..."just" do this.
    from typing import NewType
    alias_name = NewType("alias_name", alias_value)

Combine the above two approaches via The Ultimate Type Alias (TUTA),
a hidden ninja technique that supports all Python versions and static
type-checkers but may cause coworker heads to pop off like in that one
Kingsman scene:
    # Instead of this...
    from typing import TypeAlias
    alias_name: TypeAlias = alias_value

    # ..."just" do this. If you think this sucks, know that you are not alone.
    from typing import TYPE_CHECKING, NewType, TypeAlias  # <-- sus af
    from sys import version_info  # <-- code just got real
    if TYPE_CHECKING:  # <-- if static type-checking, then PEP 613
        alias_name: TypeAlias = alias_value  # <-- grimdark coding style
    elif version_info >= (3, 12):  # <-- if Python >= 3.12, then PEP 695
        exec("type alias_name = alias_value")  # <-- eldritch abomination
    else:  # <-- if Python < 3.12, then PEP 484
        alias_name = NewType("alias_name", alias_value)  # <-- coworker gives up here
```

**#JustBeartypeThings.** Feels like this should be a hashtag, if I even knew what hashtags were. :smiling_face_with_tear:
